q/verbose/verbose.q

```diff
@@ -0,0 +1,471 @@
+    ];
+    if[not type[stat] in -11 0h;
+        if[not 99h=type stat; '"invalid type for stat"];
+        if[not 11h=type key stat; '"stat must have symbol keys"];
+    ];
+    if[not -7h=type cnt; '"cnt must be long"];
+    if[not 0h=type ord; '"ord must be a general list"];
+    if[not 2=count ord; '"ord must have size 2"];
+    if[not ord[0] in (<:;>:); '"first element of ord must be <: or >:"];
+    if[not -11h=type ord[1]; '"second element of ord must be a symbol"];
+    };
+
+//wrapper to allow select in secure gateway (the ? operator has some overloads with side-effects)
+.finos.verbose.select:{[tbl;constr;grp;stat]
+    .finos.verbose.priv.validateSelectArgs[tbl;constr;grp;stat;0W;(<:;`i)];
+    ?[tbl;constr;grp;stat]};
+
+.finos.verbose.select5:{[tbl;constr;grp;stat;cnt]
+    .finos.verbose.priv.validateSelectArgs[tbl;constr;grp;stat;cnt;(<:;`i)];
+    ?[tbl;constr;grp;stat;cnt]};
+
+.finos.verbose.select6:{[tbl;constr;grp;stat;cnt;ord]
+    .finos.verbose.priv.validateSelectArgs[tbl;constr;grp;stat;cnt;ord];
+    ?[tbl;constr;grp;stat;cnt;ord]};
+
+.finos.verbose.selectPerfTest:{[tbl;constr;grp;stat]
+    if[not 99h=type stat; '"stat must be a dictionary"];
+    {[tbl;constr;grp;stcol;stexpr].Q.ts[.finos.verbose.select;(tbl;constr;grp;enlist[stcol]!enlist stexpr)][0]}[tbl;constr;grp]'[key stat;value stat]};
+
+//wrapper to allow update in secure gateway (the ! operator has some overloads with side-effects and raw update may modify tables in-place)
+.finos.verbose.update:{[tbl;constr;grp;stat]
+    if[not .Q.qt[tbl]; '".finos.verbose.update expects a table"];
+    if[not 0h=type constr; '"constraints must be a general list"];
+    if[not -1h=type grp;
+        if[not 99h=type grp; '"groupby must be boolean or dictionary"];
+        if[not 11h=type key grp; '"groupby must have symbol keys"];
+    ];
+    if[not()~stat;
+        if[not 99h=type stat; '"stat must be empty list or dictionary"];
+        if[not 11h=type key stat; '"stat must have symbol keys"];
+    ];
+    ![tbl;constr;grp;stat]};
+
+.finos.verbose.cond:{[cond;valTrue;valFalse]
+    if[not type[cond] in 1 -1h; '".finos.verbose.cond expects boolean or boolean list in the first parameter"];
+    ?[cond;valTrue;valFalse]};
+
+.finos.verbose.columnOrder: {[col;t] (distinct (col inter cols t),(cols t) except col) # t }
+
+.finos.verbose.columnRename:{[d;t]
+    d:(d?((value d) inter cols t))! (value d) inter cols t;
+    missingCols:(cols t) except value d;
+    completeDict: d,missingCols!missingCols;
+    :(completeDict ? (cols t)) # ?[t;();0b;completeDict]
+  }
+
+ //breaking this into multiple rows can break attached HDB/RDB
+.finos.verbose.safenull:{$[type[x] in 0 77h;0=count each x;87=type x;x like "";null x]};
+
+.finos.verbose.safestring:{$[type[x] in 10 87h;x;string x]};
+
+///
+// This converts list of general lists to list of a given type
+// @param t type e.g. "c" for strings
+// @param l the list
+.finos.verbose.generalListVectorToTypedVector: {[t; l]
```

Comment:
From this point on these look more like utility functions that don't align with the "provide wrappers to built-in operators" goal, but they could be moved to a file that is better suited for this.